File: 기타/201010 프로그래머스 챌린지/3번.py
def solution(A, B):
    maxAnswer = int((A+B)/4)
    maxAB = max(A, B)
    minAB = A+B-maxAB

    while maxAnswer > 0:
        if (maxAB >= maxAnswer*4) or (maxAB >= maxAnswer*3 and minAB >= maxAnswer) or (maxAB >= maxAnswer*2 and minAB >= maxAnswer*2):
            return maxAnswer
        maxAnswer -= 1

    return 0

File: 기타/210123 카카오/3번.js
function solution(next_student) {
  next_student.unshift(0);
  const visit = Array.from(Array(next_student.length), () =>
    new Array(next_student.length).fill(false)
  );
  const result = new Array(next_student.length).fill(0);

  const dfs = (current, n) => {
    if (visit[current][n]) {
      return;
    }

    visit[current][n] = true;
    result[current] += 1;

    if (next_student[n] !== 0) {
      dfs(current, next_student[n]);
    }
  };

  for (let i = 1; i < next_student.length; i++) {
    dfs(i, i);
  }

  return result.lastIndexOf(Math.max(...result));
}

const next_student = [5, 9, 13, 1, 0, 0, 11, 1, 7, 12, 9, 9, 2];
console.log(solution(next_student));
